service/tbc/peer_manager.go

```diff
@@ -5,129 +5,349 @@
-	pm.peersMtx.Unlock()
 
-	log.Debugf("PeersInsert exit %v good %v bad %v",
-		len(peers), allGoodPeers, allBadPeers)
+	pm.mtx.Lock()
+	defer pm.mtx.Unlock()
+
+	// If peer is connected don't add it to good list
+	if _, ok := pm.peers[address]; ok {
+		return fmt.Errorf("peer active: %v", address)
+	}
+
+	// Remove peer from bad.
+	delete(pm.bad, address)
+	// Add peer to good.
+	pm.good[address] = struct{}{}
+
+	log.Debugf("Good exit peers %v good %v bad %v",
+		len(pm.peers), len(pm.good), len(pm.bad))
 
 	return nil
 }
 
-// PeerDelete marks the peer as bad.
-// XXX this function only returns nil!?
-func (pm *PeerManager) PeerDelete(address string) error {
-	log.Tracef("PeerDelete")
-	defer log.Tracef("PeerDelete exit")
+func (pm *PeerManager) Connected(p *peer) {
+	log.Tracef("Connected")
+	defer log.Tracef("Connected exit")
+
+	address := p.String()
+
+	pm.mtx.Lock()
+	defer pm.mtx.Unlock()
+
+	// If peer is connected, ignore it
+	if _, ok := pm.peers[address]; !ok {
+		pm.peers[address] = p
+	}
+	delete(pm.bad, address)
+	delete(pm.good, address)
+
+	log.Debugf("Connected exit peers %v good %v bad %v",
+		len(pm.peers), len(pm.good), len(pm.bad))
+}
+
+// Bad marks the peer as bad.
+func (pm *PeerManager) Bad(address string) error {
+	log.Tracef("Bad")
+	defer log.Tracef("Bad exit")
 
 	_, _, err := net.SplitHostPort(address)
 	if err != nil {
 		return err
 	}
 
-	pm.peersMtx.Lock()
+	pm.mtx.Lock()
+
+	// If peer is connected, disconnect it and mark it bad
+	if p, ok := pm.peers[address]; ok {
+		if p != nil {
+			p.close()
+		}
+		delete(pm.peers, address)
+	}
 
 	// Remove peer from good.
-	delete(pm.peersGood, address)
+	delete(pm.good, address)
 	// Mark peer as bad.
-	pm.peersBad[address] = struct{}{}
+	pm.bad[address] = struct{}{}
 
-	// Crude hammer to reset good/bad state of peers
+	log.Debugf("Bad exit peers %v good %v bad %v",
+		len(pm.peers), len(pm.good), len(pm.bad))
 
-	// XXX goodSeenMax should be a connection test; not a threshold.
-	// Another reason to move all peer stuff into the manager.
-	pm.goodSeenMax = max(pm.goodSeenMax, len(pm.peersGood))
-	if pm.goodSeenMax > minPeersRequired && len(pm.peersGood) < minPeersRequired {
-		// Kill all peers to force caller to reseed. This happens when
-		// network is down for a while and all peers are moved into
-		// bad map.
-		clear(pm.peersGood)
-		clear(pm.peersBad)
-		pm.peersGood = make(map[string]struct{}, 8192)
-		pm.peersBad = make(map[string]struct{}, 8192)
-		pm.goodSeenMax = 0
-		log.Debugf("peer cache purged")
-	}
+	pm.mtx.Unlock()
 
-	allGoodPeers := len(pm.peersGood)
-	allBadPeers := len(pm.peersBad)
+	return nil
+}
 
-	pm.peersMtx.Unlock()
+func (pm *PeerManager) Random() (*peer, error) {
+	log.Tracef("Random")
+	defer log.Tracef("Random exit")
 
-	log.Debugf("PeerDelete exit good %v bad %v", allGoodPeers, allBadPeers)
+	pm.mtx.RLock()
+	pm.mtx.RUnlock()
+	for _, p := range pm.peers {
+		if p.isConnected() {
+			return p, nil
+		}
+	}
```

Comment:
`pm.mtx.RUnlock()` is called before reading `pm.peers`, which results in a race condition.
Did you mean to copy it before unlocking, then iterate over the copied map?
```diff
import (
+	"maps"
)
```

```suggestion
	pm.mtx.RLock()
	peers := maps.Clone(pm.peers)
	pm.mtx.RUnlock()
	for _, p := range peers {
		if p.isConnected() {
			return p, nil
		}
	}
```